src/hotspot/cpu/x86/stubGenerator_x86_64_dilithium.cpp

```diff
@@ -60,393 +57,591 @@ static address dilithiumAvx512ConstsAddr(int offset) {
+        case 128:
+          for (int i = 0; i < regCnt; i++) {
+            __ vmovdqu(output2[i], input2[i], vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evpermt2q(output2[i], unshuffle2, input1[i], vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evpermt2q(input1[i], unshuffle1, input2[i], vector_len);
+          }
+
+          break;
+        case 64:
+          for (int i = 0; i < regCnt; i++) {
+            __ vshufpd(output2[i], input1[i], input2[i], 0b11111111, vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ vshufpd(input1[i], input1[i], input2[i], 0b00000000, vector_len);
+          }
+          break;
+        case 32:
+          for (int i = 0; i < regCnt; i++) {
+            __ vmovdqu(output2[i], input2[i], vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evmovshdup(output2[i], mergeMask2, input1[i], true, vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evmovsldup(input1[i], mergeMask1, input2[i], true, vector_len);
+          }
+          break;
+        // Special cases
+        case 1: // initial shuffle for dilithiumAlmostInverseNtt
+          // shuffle all even 32bit columns to input1, and odd to input2
+          for (int i = 0; i < regCnt; i++) {
+            __ vmovdqu(output2[i], input2[i], vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evpermt2d(input2[i], unshuffle2, input1[i], vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evpermt2d(input1[i], unshuffle1, output2[i], vector_len);
+          }
+          break;
+        case 0: // final unshuffle for dilithiumAlmostNtt
+          // reverse case 1: all even are in input1 and odd in input2, put back
+          for (int i = 0; i < regCnt; i++) {
+            __ vmovdqu(output2[i], input2[i], vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evpermt2d(input2[i], unshuffle2, input1[i], vector_len);
+          }
+          for (int i = 0; i < regCnt; i++) {
+            __ evpermt2d(input1[i], unshuffle1, output2[i], vector_len);
+          }
+          break;
+        default:
+          assert(false, "Don't call here");
+      }
+    }
+  }; // return
 }
 
-// We do Montgomery multiplications of two vectors of 16 ints each in 4 steps:
+// We do Montgomery multiplications of two AVX registers in 4 steps:
 // 1. Do the multiplications of the corresponding even numbered slots into
-//    the odd numbered slots of a third register.
-// 2. Swap the even and odd numbered slots of the original input registers.
-// 3. Similar to step 1, but into a different output register.
+//    the odd numbered slots of a scratch2 register.
+// 2. Swap the even and odd numbered slots of the original input registers.*
+// 3. Similar to step 1, but into output register.
```

Comment:
Typo: into output register -> into an output register

Reply:
used 'the' to be 'specific'.. (I think the lack of articles was causing the confusion.. "the scratch2 register is combined with the output register into scratch.. or something..) Also reworded step 4?